### PR TITLE
Advanced Building Blocks - Enumerable Methods

### PR DESCRIPTION
**Pull request by afcastaneda223**
# Enumerable Methods

This Project was built with ruby, we can find these methods:

-  my_each
- my_each_with_index
- my_select 
- my_all?
- my_any?
- my_none?
- my_count
- my_map
- my_inject

### CODE REVIEW

File: Enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        if yield(x) == true
+            counter += 1
+        else
+            return n
+        end
+        }
+         counter
+    end
+    #my_map
+    return to_enum unless block_given?
```
Comment (suretrust) [inline Enumerable.rb]: This is supposed to be inside the `my_map` method, right? Let's fix this.

File: Enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    end
+    def my_all?
+        return to_enum unless block_given?
+        self.my_each { |x|
+        if yield(x) == false
+            return false
+        end
+        }
+         true
+    end
```
Comment (suretrust) [inline Enumerable.rb]: * [x] should have working `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1, true, 'hi', []].my_all? != [1, true, 'hi', []].all?

[1,2,3,4,5,6].my_all?(Integer) != [1,2,3,4,5,6].all?(Integer)

['saheed', 'suretrust', 'oladele'].my_all?(/d/) != ['saheed', 'suretrust', 'oladele'].all?(/d/)

[1,2,3,4,5,6].my_all?(3) != [1,2,3,4,5,6].all?(3)
```

File: Enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    #my_any?
+    def my_any?
+        return to_enum unless block_given?
+        self.my_each { |x|
+        if yield(x) == true
+            return true
+        end
+        }
+         false
+    end
```
Comment (suretrust) [inline Enumerable.rb]: * [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[nil, false, true, []].my_any? != [nil, false, true, []].any?

[1,2,3,4,5,6].my_any?(Integer) != [1,2,3,4,5,6].any?(Integer)

['saheed', 'suretrust', 'oladele'].my_any?(/z/) != ['saheed', 'suretrust', 'oladele'].any?(/z/)

['saheed', 'suretrust', 'oladele'].my_any?('suretrust') != ['saheed', 'suretrust', 'oladele'].my_any?('suretrust')
```

File: Enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    #my_none?
+    def my_none?
+        return to_enum unless block_given?
+        self.my_each { |x|
+        if yield(x) == true
+            return false
+        end
+        }
+         true
+    end
```
Comment (suretrust) [inline Enumerable.rb]: * [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[nil, false, nil, false].my_none? != [nil, false, nil, false].none?

[1,2,3,4,5,6,7].my_none?(String) != [1,2,3,4,5,6,7].none?(String)

['saheed', 'suretrust', 'oladele'].my_none?(/z/) != ['saheed', 'suretrust', 'oladele'].none?(/z/)

['saheed', 'suretrust', 'oladele'].my_none?(5) != ['saheed', 'suretrust', 'oladele'].none?(5)
```

File: Enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        n= self.length
+        self.my_each { |x|
+        if yield(x) == true
+            counter += 1
+        else
+            return n
+        end
+        }
+         counter
+    end
```
Comment (suretrust) [inline Enumerable.rb]: * [ ] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5,6,7].my_count != [1,2,3,4,5,6,7].count

[1,2,3,4,2,5,6,7].my_count(2) != [1,2,3,4,2,5,6,7].count(2)
```

File: Enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        return to_enum unless block_given?
+        n= self.length
+        x = 1
+        total = self[0]
+        while x < n
+           total = yield(total,self[x])
+            x += 1
+        end
+        return total
+    end
```
Comment (suretrust) [inline Enumerable.rb]: * [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
Range.new(5, 50).my_inject(4) { |prod, n| prod * n } => gives error

[1,2,3,4,5,6,7].my_inject(:+) != [1,2,3,4,5,6,7].inject(:+)
```

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @afcastaneda223, 👋 

### Great job thus far. 🙌 Please see my comments. 👇 

- [ ] Add linter to your repo.

- [ ] Fix all suggestions below.

**Kindly submit for another code review request once these issues have been fixed.** _Happy coding!_ 🎉

File: Enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+    #my_each_with_index
+    def my_each_with_index
+        return to_enum unless block_given?
+        x=0
+        self.my_each{
+            yield(self[x], x)
+            x += 1
+        }
+        end
```
Comment (maelfosso) [inline Enumerable.rb]: Not needed. It's the caused of the error.

File: Enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+                return false unless x.is_a?(arg)
+               }
+           return true
+            else arg != nil
+                self.my_each{ |x|
+                return false unless x == arg
+                }
+            return true
+        end
+    end
```
Comment (maelfosso) [inline Enumerable.rb]:
```rb
2.6.5 :016 > true_array = [1, true, 'hi', []]
2.6.5 :017 > true_array.all?
 => true 
2.6.5 :018 > true_array.my_all?
 => false 
2.6.5 :019 >
```

File: Enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+
+            else arg !=nil
+                self.my_each{ |x|
+                if x == arg
+                    return true 
+                end
+                }
+            return false
+        end
+    end
```
Comment (maelfosso) [inline Enumerable.rb]:
```rb
2.6.5 :020 > true_array = [1, true, 'hi', []]
2.6.5 :021 > true_array.my_any?
 => false 
2.6.5 :022 > true_array.any?
 => true 
2.6.5 :023 >
```

File: Enumerable.rb [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+
+            else arg !=nil
+                self.my_each{ |x|
+                if x == arg
+                    return false 
+                end
+                }
+                return true
+        end
+    end
```
Comment (maelfosso) [inline Enumerable.rb]:
```rb
2.6.5 :026 > true_array = [1, true, 'hi', []]
2.6.5 :027 > true_array.my_none?
 => true 
2.6.5 :028 > true_array.none?
 => false 
2.6.5 :029 >
```

File: Enumerable.rb [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+           total = yield(total,self[x])
+            x += 1
+        end
+        return total
+    end
+
+    def multiply_els
+        self.my_inject{|x,y| x*y}
+    end
+end
```
Comment (maelfosso) [inline Enumerable.rb]:
```rb
2.6.5 :042 > range = Range.new(5, 50)
2.6.5 :043 > actual = range.my_inject(4) { |prod, n| prod * n }
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):43
        1: from Enumerable.rb:184:in `my_inject'
ArgumentError (wrong number of arguments (given 1, expected 0))
2.6.5 :044 > expected = range.inject(4) { |prod, n| prod * n }
2.6.5 :045 >
```

Comment (maelfosso) [reply Enumerable.rb]:
```rb
2.6.5 :047 > array
 => [3, 8, 5, 7, 1, 6, 1, 4, 6, 3, 8, 8, 2, 0, 6, 3, 6, 4, 4, 4, 5, 6, 4, 7, 5, 1, 2, 5, 1, 1, 2, 2, 7, 6, 1, 8, 0, 4, 6, 5, 1, 4, 2, 8, 2, 7, 0, 7, 6, 6, 6, 4, 3, 0, 4, 8, 6, 2, 3, 4, 6, 5, 3, 4, 2, 4, 8, 7, 1, 3, 1, 4, 2, 8, 7, 3, 7, 5, 5, 2, 8, 4, 3, 0, 3, 0, 3, 7, 6, 0, 7, 4, 5, 2, 1, 2, 1, 1, 4, 5] 
2.6.5 :048 > array.inject(:+)
 => 403 
2.6.5 :049 > array.my_inject(:+)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):49
        1: from Enumerable.rb:184:in `my_inject'
ArgumentError (wrong number of arguments (given 1, expected 0))
2.6.5 :050 > 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thanks for applying the suggested changes.
However, there are some warning and bug into your code

1- I check the syntax of your code
```rb
$ ruby -wc Enumerable.rb 
Enumerable.rb:21: warning: mismatched indentations at 'end' with 'def' at 14
Enumerable.rb:23: warning: mismatched indentations at 'end' with 'module' at 1
Enumerable.rb:45: warning: mismatched indentations at 'elsif' with 'if' at 37
Enumerable.rb:51: warning: mismatched indentations at 'elsif' with 'if' at 37
Enumerable.rb:57: warning: mismatched indentations at 'elsif' with 'if' at 37
Enumerable.rb:62: warning: mismatched indentations at 'else' with 'if' at 37
Enumerable.rb:62: warning: possibly useless use of != in void context
Enumerable.rb:78: warning: mismatched indentations at 'elsif' with 'if' at 71
Enumerable.rb:85: warning: mismatched indentations at 'elsif' with 'if' at 71
Enumerable.rb:92: warning: mismatched indentations at 'elsif' with 'if' at 71
Enumerable.rb:100: warning: mismatched indentations at 'else' with 'if' at 71
Enumerable.rb:100: warning: possibly useless use of != in void context
Enumerable.rb:118: warning: mismatched indentations at 'elsif' with 'if' at 111
Enumerable.rb:125: warning: mismatched indentations at 'elsif' with 'if' at 111
Enumerable.rb:132: warning: mismatched indentations at 'elsif' with 'if' at 111
Enumerable.rb:140: warning: mismatched indentations at 'else' with 'if' at 111
Enumerable.rb:140: warning: possibly useless use of != in void context
Enumerable.rb:158: warning: mismatched indentations at 'elsif' with 'if' at 152
Enumerable.rb:164: warning: mismatched indentations at 'else' with 'if' at 152
Enumerable.rb:175: warning: mismatched indentations at 'elsif' with 'if' at 173
Enumerable.rb:177: warning: mismatched indentations at 'else' with 'if' at 173
Enumerable.rb:198: syntax error, unexpected end, expecting end-of-input
```
Fix it to move forward

2- Read the comments left below and correct the suggested changes

3- Your PR is not following the GitFlow. Your must create the PR from a feature branch, requesting a merging into the development branch. Correct that, please.

Kindly make the suggested changes and submit a new code review
Thank

Mael FOSSO

**Review by trekab — CHANGES_REQUESTED** [review]
## Changes Required! :recycle: 

### my_each:
- [ ] when a block is given when self is a range, it should return the range itself after calling the given block once for each element in self
- [ ] when a block is given when self is a range, it should behave like `each` i.e. it calls the given block once for each element in self
- [ ] when a block is given when self is a hash, it should behave like `each` i.e. it calls the given block once for each element in self
### my_each_with_index:
- [ ] when a block is given when self is a range, it should return the range itself after calling the given block once for each element in self
- [ ] when a block is given when self is a range, it should behave like `each_with_index` i..e it calls the given block once for each element in self
- [ ] when a block is given when self is a hash, it should behave like `each_with_index` i.e. it calls the given block once for each element in self
### my_inject:
- [ ] should raise a "LocalJumpError" when no block or argument is given
- [ ] when a symbol is specified without an initial value, should combine each element of the collection by applying the symbol as a named method.
- [ ] when a symbol is specified with an initial value, should combine each element of the collection by applying the symbol as a named method.
### Others:
- [ ] `my_select` should return an array containing all elements of enum for which the given block returns a true value. [Compare docs](https://ruby-doc.org/core-2.6.5/Enumerable.html#method-i-select)
- [ ] `my_all?` should return false if the block returns false or nil.
- [ ] `my_any?` should return false if the block ever returns a value false or nil.
- [ ] `my_none?` should return false if the block ever returns true for all elements.
- [ ] `my_count` should return the number of items in enum through enumeration.
- [ ] `my_map` should return a new array with the results of running a given block once for every element in the enum.
- [ ] `multiply_els` should accept an array as an argument and multiply all the elements of the array together using `my_inject`
- [ ] Please try to follow the correct GitFlow or GitHub flow on your next project.

__P.S:__ Please feel free to compare your methods with those in this [documentation](https://ruby-doc.org/core-2.6.5/Enumerable.html) and don't hesitate to reach out for any further clarification.

__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink: 
__Treasure Kabareebe__
[GitHub](https://github.com/trekab) | [LinkedIn](https://www.linkedin.com/in/treasure-kabareebe/) | [Twitter](https://twitter.com/TKabareebe)

**Comment by sumancrest0001** [comment]
**Status: Changes required 
Hello @afcastaneda223,**
We are progressing nicely, I know this project is quite difficult as we have to make all the enumerable methods to work similar to the default enumerable methods. You are doing well but we still have some issues that need to be addressed before moving to the next project. Let's have a look at the following failed tests and make changes accordingly.

**#my_each**
- when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- when a block is given when self is a range behaves like #each calls the given block once for each element in self
- when a block is given when self is a hash behaves like #each calls the given block once for each element in self

**#my_each_with_index** 
- when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
- when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

**#my_select** returns an array containing all elements of enum for which the given block returns a true value::range
**#my_all?** returns false if the block returns false or nil::range
**#my_any?** returns false if the block ever returns a value false or nil::range
**#my_none?** returns false if the block ever returns true for all elements::range
**#my_count** returns the number of items in enum through enumeration::range
**#my_map** returns a new array with the results of running a given block once for every element in enum::range

**#my_inject**
- when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
- when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range

**Review by sumancrest0001 — CHANGES_REQUESTED** [review]
You can reference [this](https://ruby-doc.org/core-2.6.5/Enumerable.html) documentation to know more about enumerable methods and examples.

If you need any help, message me on slack **Suman Shrestha**. I am ready to help.

**Review by sumancrest0001 — APPROVED** [review]
**Status: Approved :+1: :100: 
Congratulation @afcastaneda223  :clap:** 
Your custom methods are passing all the tests. It is approved. Now, you can merge the PR to master. Before that, please follow the following suggestions to make your repository professional.
[Important] Please follow git-flow while working on the project. If you do not follow it while working on a project, it will be invalid.[ reference](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/git_flow.png)
[Optional] Remove all the unnecessary files and code to make your repository clean. It is not necessary to push test1 and test2 files. we can also remove the test samples you are using in the Enumerable.rb file.
[Optional] Describe the project on the readme file and also add steps to run your project on a local machine.

**Pull request merged**